apps/extension/src/offscreen.ts

```diff
@@ -41,30 +45,31 @@ export const buildActionHandler = (request: ActionBuildRequest) => {
   // Destructure the data object to get individual fields
   const { transactionPlan, witness, fullViewingKey } = request;
 
-  return transactionPlan.actions.map((_, i) =>
-    spawnWorker(transactionPlan, witness, fullViewingKey, i),
+  return transactionPlan.actions.map((_, actionPlanIndex) =>
+    spawnWorker({ transactionPlan, witness, fullViewingKey, actionPlanIndex }),
   );
 };
 
-const spawnWorker = (
-  transactionPlan: JsonObject & { actions: JsonValue[] },
-  witness: JsonObject,
-  fullViewingKey: string,
-  actionPlanIndex: number,
-): Promise<JsonValue> => {
+const spawnWorker = ({
+  transactionPlan,
+  witness,
+  fullViewingKey,
+  actionPlanIndex,
+}: WasmBuildActionInput): Promise<WasmBuildActionOutput> => {
   return new Promise((resolve, reject) => {
     const worker = new Worker(new URL('./wasm-build-action.ts', import.meta.url));
 
-    const onWorkerMessage = (e: MessageEvent) => {
-      resolve(e.data as JsonValue);
+    const onWorkerMessage = (e: MessageEvent<unknown>) => {
       worker.removeEventListener('error', onWorkerError);
       worker.terminate();
+      if (isJsonObject(e.data)) resolve(e.data);
+      else reject(new Error(`No Action data from worker ${actionPlanIndex}`));
     };
 
     const onWorkerError = (error: ErrorEvent) => {
-      reject(new Error(`${error.filename}:${error.lineno} ${error.message}`));
       worker.removeEventListener('message', onWorkerMessage);
       worker.terminate();
+      reject(new Error(`${error.filename}:${error.lineno} ${error.message}`));
```

Comment:
I suppose we can't `reject(error)`?

Reply:
the 'error' is actually an `ErrorEvent`, not an `Error`. i can update the name to make that clear